### ISSUE TITLE
Support file language detection by extension code

### ISSUE THREAD

**Issue by TonyGravagno** [microsoft/vscode#80303]
We've seen requests here on the topic of extension/plugin determination of language modes. They all seem to come back to a focus on known markers, whether file extensions, first lines or a similar premble, or the value of a schema spec. Ref other topics #10915, #11838, #13782, #73424, #53350, #69868. All of these discussions include the notion of an open API that allows an extension to determine the language for a file. But they all seem to conclude with a focus on specific languages, whether JS, JSON, PHP, or XML implementations, and the idea that new options will need to conform to existing processes.

I'm a member of a non-relational DBMS industry that's been around for over 50 years, using several dialects of the original Dartmouth BASIC. Developers in this industry are slowly adopting VSCode for maintaining code and other documents and features of the environment. Pre-dating DOS and Unix, we have never required extensions on filenames. The first line of a program can be any valid statement, so using line 1 as a marker isn't valid either.

I recommend merging all of the related repo issues and resolving them with a single, simple change:
1) Run the routine that processes file.associations, as usual.
2) Invoke a new hook which gives a registered extension an opportunity to reset the determined language. Might we be talking about a single line of code in the VSCode core?
3) Then continue as normal to invoke the module that's registered to handle the language.

The extension chosen by the user will be FOSS, maintained, installed, and replaced like any other. No one will own the process for language mode selection anymore, and all of these discussions about new ways to choose the language mode go away.

Like editor.defaultFormatter, there can only be one of these extensions registered. For users with a need for multiple extensions that determine the language mode (last one wins), a mechanism can be designed in the field to allow a cascade of handling, similar to the way linters process sets of rules. Or one extension might attempt to handle all such logic. Again, this is a detail to be shaken out in the field.

If required, I'll get into the code and attempt to submit a PR, but since others here are already familiar with this code, and since this one change will resolve a number of issues, I'm hoping that won't be necessary. :)

In summary, please don't hardcode rules as a way of providing more freedom to specific groups. We've all recognized the pattern in the referenced requests, and the proposals so far just defer the problem until a new request comes in. The solution to accommodating a wide variety of requests is to call out to external, user-provided functions that make choices which aren't possible with a general-purpose tool. Thanks.

**Comment by jrieken** [comment]
> Invoke a new hook which gives a registered extension an opportunity to reset the determined language. 

Like this: [`vscode.languages.setTextDocumentLanguage`](https://github.com/microsoft/vscode/blob/bd79c20ac85d44a42d28defb6aa280d5e9ab066b/src/vs/vscode.d.ts#L8088)?

**Comment by TonyGravagno** [comment]
Thanks Johannes. While that's one part of it, that's not what I'm thinking.

I have an extension that provides language features, but I wouldn't want to be a bad citizen and activate it all the time. To limit the activation scope of the extension, for specific projects, we can use `"activationEvents": ["workspaceContains:**/.foo"]`. But that won't help when one-off documents are loaded.

So I'm thinking of loading a much smaller extension (I'll call this MiniMe) with `"activationEvents": [ "*"]` and have it load the bigger extension when required. I believe I need to handle onDidChangeActiveTextEditor, get the active document, and if it has no languageId, parse for language markers to see if the bigger extension is required, and then use setTextDocumentLanguage if required.

I'm asking if there is a better place to put MiniMe code like that, some already available, ever-present module where we can put a few small functions? I don't know if we can subscribe to events from a Task which might be invoked on startup. Or is it better protocol to create a MiniMe for every application like this? I can imagine a single developer wanting to do this for a couple languages that aren't auto-detected - why have more than one MiniMe, each one loading its own larger extension, when one such module can handle several?

Code like this might also want to take action only if there is a failure to identify the language for the currently opened document. So, there is `"activationEvents": [ "onLanguage:foo"]` but is there an activation event when no language has been selected?

And to my original thoughts - Is this the way we all should be handling this common desire for handling language selection without using the file extension?

**Comment by jrieken** [comment]
> So, there is "activationEvents": [ "onLanguage:foo"] but is there an activation event when no language has been selected?

The default is `plaintext` so when we fail to identify a language it will be `onLanguage:plaintext`. You could use that but there is potential for false positives (which I believe you cannot avoid anyways given the lack of a file extension)

> I'm asking if there is a better place to put MiniMe code like that, some already available, ever-present module where we can put a few small functions? 

We split between declarative data (e.g JSON declared things like activation events etc) and code. That's a nice and clean cut and allows to delay code loading (which often is costly). I like the MiniMe approach and kudos for wanting to be a nice citizen! Formalising the MiniMe-approach is hard tho, how little code is still Mini? Single file? Less than N lines of code? No imports? That's why we like the cut between declarative data and executable source - no such questions.

For your setup I would measure? How big is your extension and how long does it take to load (F1 > Startup Performance > find "Extension Activation Stats")? Can you load the heavy bits lazy (`await import('./heavy.js')`) (e.g that a similar to approach MiniMe and MaxMe extensions using a single extension and multiple JS bundles)

> Is this the way we all should be handling this common desire for handling language selection without using the file extension?

This issues hasn't come up much TBH. In the [languages contrib point](https://code.visualstudio.com/api/references/contribution-points#contributes.languages) we support extensions, filename pattern, and firstLine-regexp which has been sufficient till now. An option would be to allow not only for firstLine-regexp but also for "allLines"-regexp but that I believe that's a slippery slope, blurring the line between declarative and executable data and then you might be better of with activating a lean extension that makes the decision.

**Comment by aeschli** [comment]
I agree with @jrieken that we need to do this declaratively. Maybe we need more than what we have now such as https://github.com/microsoft/vscode/issues/69868.
